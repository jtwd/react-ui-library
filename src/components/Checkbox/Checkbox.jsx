import React from 'react'
import { string, func, bool } from 'prop-types'
import { compose, withProps } from 'recompose'

import { IconWrapper, StyledInput, StyledLabel } from "./Checkbox.styles"
import Icon from '../Icon'
import withFocus from '../_enhancers/withFocus'
import withChecked from '../_enhancers/withChecked'

function Checkbox({label,htmlId, isChecked, toggleChecked, hasFocus, toggleFocus, iconName, value}) {
  return (
    <StyledLabel htmlFor={htmlId} checked={isChecked} focus={hasFocus}>
      <StyledInput
        type="checkbox"
        id={htmlId}
        value={value}
        checked={isChecked}
        onChange={toggleChecked}
        onFocus={toggleFocus}
        onBlur={toggleFocus}
      />
      <IconWrapper>
        <Icon icon={iconName} />
      </IconWrapper>
      {label}
    </StyledLabel>
  )
}

Checkbox.propTypes = {
  /** Text to be displayed in label and value (if no value prop is supplied) */
  label: string.isRequired,
  /** to be called with checkbox changes */
  toggleChecked: func.isRequired,
  /** Unique id */
  htmlId: string.isRequired,
  /** Value prop (label is used if this is not supplied) */
  value: string.isRequired,
  /** Checked value */
  isChecked: bool,
  /** Focus value */
  hasFocus: bool.isRequired,
  /** Function to toggle focus */
  toggleFocus: func.isRequired,
  /** Name of icon to display */
  iconName: string.isRequired
}

Checkbox.defaultProps = {
  isChecked: false
}

const enhance = compose(
  withFocus,
  withChecked,
  withProps(({ isChecked, value, label }) => {
    // if value is not provided, use label
    let newValue = label
    if (value) newValue = value
    return {
      iconName: isChecked ? 'checkboxOn' : 'checkboxOff',
      value: newValue
    }
  })
)

export default enhance(Checkbox)

/*
export default compose(
  withFocus,
  withChecked,
  withProps(({ isChecked, value, label }) => {
    // if value is not provided, use label
    let newValue = label
    if (value) newValue = value
    return {
      iconName: isChecked ? 'checkboxOn' : 'checkboxOff',
      value: newValue
    }
  })
)(Checkbox)
*/