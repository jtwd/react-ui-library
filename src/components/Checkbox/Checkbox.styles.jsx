import styled from 'styled-components'

import { transBlack, transBlackLight, transBlackDark, transWhiteDark, yellow, white, darkYellow } from "../_theme/colors"
import { radiusRound, transitions } from "../_theme/units"
import { xxs, xs, xxl } from "../_theme/spacers"
import trimChildren from '../_theme/mixins/trimChildren'

const hoverFocusStyles = `
  background: ${transBlack};
  color: ${white};
`

export const StyledLabel = styled.label`
  position: relative;
  padding: ${xxs} ${xxs} ${xxs} 2.5rem;
  display: block;
  cursor: pointer;
  border-radius: ${radiusRound};
  background: ${transBlackLight};
  transition: ${transitions.hover};
  color: ${transWhiteDark};
  margin-bottom: ${xxs};
  ${trimChildren('ver')}
  
  &:hover {
    ${hoverFocusStyles}
  }
  
  ${props => props.focus && `
    ${hoverFocusStyles}
  `}
  
  ${props => props.checked && `
    color: ${yellow};
    background: ${transBlack};
    &:hover, &:focus {
      color: ${darkYellow};
    }
  `}
  ${props => props.checked && props.focus && `
    color: ${darkYellow};
    background: ${transBlackDark};
  `}
`

export const IconWrapper = styled.span`
  display: inline-block;
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
  margin-top: 0.225rem;
  left: ${xs};
`

export const StyledCheckbox = styled.input`
  position: absolute;
  z-index: -1;
  opacity: 0;
`
