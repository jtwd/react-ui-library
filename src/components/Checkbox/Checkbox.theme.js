import { getTheme } from '../_theme/utils'
import * as c from '../_theme/palette'

const { spacing, radius, transitions } = getTheme()

export default {
  default: {
    padding: `${spacing.xxs} ${spacing.xxs} ${spacing.xxs} 2.5rem`,
    borderRadius: radius.rounded,
    backgroundColor: c.trans(1),
    color: c.trans(4, false),
    hoverBackgroundColor: c.trans(),
    hoverColor: c.light(),
    checkedBackgroundColor: c.trans(),
    checkedColor: c.primary(),
    checkedHoverColor: c.primary('dark'),
    checkedHoverBackgroundColor: c.trans(4),
    transition: transitions.hover,
    margin: `0 0 ${spacing.xxs}`,
    iconTop: spacing.xxs,
    iconLeft: spacing.xs
  }
}