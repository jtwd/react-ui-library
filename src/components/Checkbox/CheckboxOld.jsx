import React, { Component } from 'react'
import { string, func, bool } from 'prop-types'
import { IconWrapper, StyledInput, StyledLabel } from "./Checkbox.styles"
import Icon from '../Icon'

class Checkbox extends Component {
  constructor(props) {
    super(props)
    this.state = {
      isChecked: false,
      hasFocus: false
    }
  }

  componentWillMount () {
    if (this.props.checked) this.setState(({isChecked}) => ({ isChecked: !isChecked }))
  }

  toggleCheckboxChange() {
    const { handleCheckboxChange, label } = this.props
    this.setState(({isChecked}) => ({ isChecked: !isChecked }))
    handleCheckboxChange(label)
  }

  toggleFocus () {
    this.setState(({hasFocus}) => ({ hasFocus: !hasFocus }))
  }

  render () {
    const { label, htmlId, value } = this.props
    const { isChecked, hasFocus } = this.state
    const checkboxIconName = isChecked ? 'checkboxOn' : 'checkboxOff'

    let chkValue = label
    if (value) chkValue = value

    return (
      <StyledLabel htmlFor={htmlId} checked={isChecked} focus={hasFocus}>
        <StyledInput
          type="checkbox"
          id={htmlId}
          value={chkValue}
          checked={isChecked}
          onChange={() => this.toggleCheckboxChange()}
          onFocus={() => this.toggleFocus()}
          onBlur={() => this.toggleFocus()}
        />
        <IconWrapper>
          <Icon icon={checkboxIconName} />
        </IconWrapper>
        {label}
      </StyledLabel>
    )
  }
}

Checkbox.propTypes = {
  /** Text to be displayed in label and value (if no value prop is supplied) */
  label: string.isRequired,
  /** to be called with checkbox changes */
  handleCheckboxChange: func.isRequired,
  /** Unique id */
  htmlId: string.isRequired,
  /** Value prop (label is used if this is not supplied) */
  value: string,
  /** Checked value */
  checked: bool
}

Checkbox.defaultProps = {
  checked: false,
  value: null
}

export default Checkbox
