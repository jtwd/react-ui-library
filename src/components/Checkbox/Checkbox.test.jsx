import React from 'react'
import {shallow} from 'enzyme'
import 'jest-styled-components'

import Checkbox from './Checkbox'

describe('Checkbox component', () => {
  it('renders a checkbox by default', () => {
    const props = {
      htmlId: 'chkbox1',
      label: 'Default checkbox',
      value: 'yes'
    }
    const component = shallow(
      <Checkbox {...props} />
    )
    expect(component).toMatchSnapshot()
  })
  it('renders a checked checkbox when given the isChecked prop', () => {
    const props = {
      htmlId: 'chkbox1',
      label: 'Default checkbox',
      value: 'yes',
      isChecked: true
    }
    const component = shallow(
      <Checkbox {...props} />
    )
    expect(component).toMatchSnapshot()
  })
})