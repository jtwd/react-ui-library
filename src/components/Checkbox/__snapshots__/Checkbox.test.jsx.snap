// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox component renders a checkbox by default 1`] = `
<withHandlers(withState(withHandlers(withProps(Checkbox))))
  hasFocus={false}
  htmlId="chkbox1"
  label="Default checkbox"
  toggleFoc={[Function]}
  value="yes"
/>
`;

exports[`Checkbox component renders a checked checkbox when given the isChecked prop 1`] = `
<withHandlers(withState(withHandlers(withProps(Checkbox))))
  hasFocus={false}
  htmlId="chkbox1"
  isChecked={true}
  label="Default checkbox"
  toggleFoc={[Function]}
  value="yes"
/>
`;
